src/core/server/metrics/metrics_service.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ReplaySubject } from 'rxjs';
+import { first, shareReplay } from 'rxjs/operators';
+import { CoreService } from '../../types';
+import { CoreContext } from '../core_context';
+import { Logger } from '../logging';
+import { InternalHttpServiceSetup } from '../http';
+import { InternalMetricsServiceSetup, InternalMetricsServiceStart, OpsMetrics } from './types';
+import { OpsMetricsCollector } from './ops_metrics_collector';
+import { opsConfig, OpsConfigType } from './ops_config';
+
+interface MetricsServiceSetupDeps {
+  http: InternalHttpServiceSetup;
+}
+
+/** @internal */
+export class MetricsService
+  implements CoreService<InternalMetricsServiceSetup, InternalMetricsServiceStart> {
+  private readonly logger: Logger;
+  private metricsCollector?: OpsMetricsCollector;
+  private collectInterval?: NodeJS.Timeout;
+  private metrics$ = new ReplaySubject<OpsMetrics>(1);
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.logger = coreContext.logger.get('metrics');
+  }
+
+  public async setup({ http }: MetricsServiceSetupDeps): Promise<InternalMetricsServiceSetup> {
+    this.metricsCollector = new OpsMetricsCollector(http.server);
+
+    const metricsObservable = this.metrics$.pipe(shareReplay(1));
+
+    return {
+      getOpsMetrics$: () => metricsObservable,
+    };
+  }
+
+  public async start(): Promise<InternalMetricsServiceStart> {
+    if (!this.metricsCollector) {
+      throw new Error('#setup() needs to be run first');
+    }
+    const config = await this.coreContext.configService
+      .atPath<OpsConfigType>(opsConfig.path)
+      .pipe(first())
+      .toPromise();
+
+    await this.refreshMetrics();
+
+    this.collectInterval = setInterval(() => {
```

Comment:
Should we collect metrics if there are no subscribers? OTOH as I understand server metrics are stateful so a lazy subscription might affect them.

Reply:
You are right, server metrics are stateful, the server collector must listen to the lifecycle events even when `collect` is not called, and this probably represents most of the collectors cpu/mem usage. 

We could still use rxjs interval instead of setInterval to avoid collecting the metrics when no subscribers are present, however this complexify the code a little (well, at least the tests, as it's a pain to test rxjs intervals), and (effectively), there will always be subscribers as at least the `status` plugin is going to consume the API. This, the default interval being quite huge, and the fact that the actual node/system calls used are very low perf consumption makes me think this should be alright.  wdyt?

Reply:
> there will always be subscribers as at least the status plugin is going to consume the API.

then the current implementation is fine